### PR TITLE
cg_api: implement `trap_R_BatchInPVS` to test if a bunch of entities are in PVS in one call

### PR DESCRIPTION
**Pull request by illwieckz**
Implement `trap_R_inPVSArray` to test array of entities in PVS in one call.

Engine sidecar of:

- https://github.com/Unvanquished/Unvanquished/pull/2679

### CODE REVIEW

**Comment by illwieckz** [comment]
I rewrote this using lessons learned while implementing:

- https://github.com/DaemonEngine/Daemon/pull/1114/files

The code now sends and receives the shortest vector possible for valid and active entities only.

File: src/shared/client/cg_api.cpp [hunk 1]
```diff
@@ -460,6 +460,17 @@ bool trap_R_inPVS( const vec3_t p1, const vec3_t p2 )
 	return res;
 }
 
+void trap_R_BatchInPVS(
+	const vec3_t origin,
+	const std::vector<std::array<float, 3>>& posEntities,
+	std::vector<bool>& inPVS )
+{
+	std::array<float, 3> myOrigin;
+	VectorCopy(origin, myOrigin.data());
```
Comment (slipher) [inline src/shared/client/cg_api.cpp:469]: The `.data` is not needed.

**Review by slipher — COMMENTED** [review]
We should ultimately avoid the IPC entirely by implementing inPVS in the cgame, but this is fine for now.

LGTM to the engine side.

File: src/engine/client/cl_cgame.cpp [hunk 2]
```diff
@@ -1248,6 +1248,25 @@ void CGameVM::QVMSyscall(int syscallNum, Util::Reader& reader, IPC::Channel& cha
...
+		case CG_R_BATCHINPVS:
+			IPC::HandleMsg<Render::BatchInPVSMsg>(channel, std::move(reader), [this] (
+				const std::array<float, 3>& origin,
+				const std::vector<std::array<float, 3>>& posEntities,
+				std::vector<bool>& inPVS)
+			{
+
+				size_t numEntities = posEntities.size();
+
+				inPVS.reserve(MAX_LOCATIONS);
```
Comment (illwieckz) [inline src/engine/client/cl_cgame.cpp:1260]: Does it makes sense to also do a `inPVS.reserve(MAX_LOCATIONS);` here as `inPVS` is an argument?

I tried it, the code builds and run, but I don't know if that changes something.

File: src/shared/client/cg_api.cpp (hunk 1 above)
Comment (illwieckz) [reply src/shared/client/cg_api.cpp]: Removed.

File: src/engine/client/cl_cgame.cpp (hunk 2 above)
Comment (slipher) [reply src/engine/client/cl_cgame.cpp:1260]: It doesn't make sense to have both `reserve` and `resize`. You can do either reserve+push_back, or resize+`a[b]=c`.

Comment (illwieckz) [reply src/engine/client/cl_cgame.cpp]: I rewrote that part, in fact I could iterate the source array directly by using `push_back` for the destination, which is cleaner.

File: src/engine/client/cg_api.h [hunk 3]
```diff
@@ -163,6 +163,12 @@ bool        trap_GetEntityToken( char *buffer, int bufferSize );
 std::vector<std::vector<Keyboard::Key>> trap_Key_GetKeysForBinds(int team, const std::vector<std::string>& binds);
 int             trap_Key_GetCharForScancode( int scancode );
 bool        trap_R_inPVS( const vec3_t p1, const vec3_t p2 );
+
+void trap_R_BatchInPVS(
+	const vec3_t origin,
+	const std::vector<std::array<float, 3>>& posEntities,
+	std::vector<bool>& inPVS );
```
Comment (slipher) [inline src/engine/client/cg_api.h:170]: How about returning the result?

File: src/engine/client/cl_cgame.cpp [hunk 4]
```diff
@@ -1287,6 +1287,21 @@ void CGameVM::QVMSyscall(int syscallNum, Util::Reader& reader, IPC::Channel& cha
 			});
 			break;
 
+		case CG_R_BATCHINPVS:
+			IPC::HandleMsg<Render::BatchInPVSMsg>(channel, std::move(reader), [this] (
+				const std::array<float, 3>& origin,
+				const std::vector<std::array<float, 3>>& posEntities,
+				std::vector<bool>& inPVS)
+			{
+				inPVS.resize(posEntities.size());
```
Comment (slipher) [inline src/engine/client/cl_cgame.cpp:1296]: Did you mean `reserve`?

Comment (illwieckz) [reply src/engine/client/cl_cgame.cpp]: Right that's what I meant. Fixed.

What I'm not sure is, how do we are sure the size is always `0` here?

File: src/engine/client/cg_api.h (hunk 3 above)
Comment (illwieckz) [reply src/engine/client/cg_api.h:170]: What do you mean?

Comment (slipher) [reply src/engine/client/cg_api.h:170]: Return the `vector<bool>` instead of writing into a reference.

Comment (illwieckz) [reply src/engine/client/cg_api.h:170]: Wouldn't this produce an extra copy of the array?

Comment (slipher) [reply src/engine/client/cg_api.h:170]: No. A return value can be automatically moved from a local variable.

File: src/engine/client/cl_cgame.cpp (hunk 4 above)
Comment (slipher) [reply src/engine/client/cl_cgame.cpp]: The IPC system will construct an empty vector.

Comment (illwieckz) [reply src/engine/client/cl_cgame.cpp]: OK! That's good to know.

File: src/shared/client/cg_api.cpp [hunk 5]
```diff
@@ -472,6 +472,17 @@ bool trap_R_inPVS( const vec3_t p1, const vec3_t p2 )
 	return res;
 }
 
+std::vector<bool> trap_R_BatchInPVS(
+	const vec3_t origin,
+	const std::vector<std::array<float, 3>>& posEntities )
+{
+	std::array<float, 3> myOrigin;
+	VectorCopy(origin, myOrigin.data());
```
Comment (slipher) [inline src/shared/client/cg_api.cpp:480]: Nit: `.data()` is not needed

**Review by slipher — APPROVED** [review]
LGTM

File: src/shared/client/cg_api.cpp (hunk 5 above)
Comment (illwieckz) [reply src/shared/client/cg_api.cpp:480]: Right, this was mimicking what other functions do, I'll remove that and maybe we can do further PR to simplify code of other functions.

**Comment by illwieckz** [comment]
> LGTM

Any comment on the game counterpart?

- https://github.com/Unvanquished/Unvanquished/pull/2679

**Pull request merged**